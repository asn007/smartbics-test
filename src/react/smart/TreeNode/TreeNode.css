:local .tree, :local .tree ul {
  list-style-type: none;
  padding: 0;
  margin: 0 0 0 1rem;
  position: relative;
}

:local .tree ul {
  margin-left: 0.5rem;
}

:local .tree:before, :local .tree ul:before {
  content: "";
  position: absolute;
  display:block;
  top: 0;
  left: 0;
  bottom: 0;
  width: 0;
  border-left: 1px solid;
}


:local .tree li {
  margin: 0;
  padding: 0 1.5rem;
  line-height: 2rem;
  position: relative;
}


:local .tree li:before {
  content:"";
  display:block;
  width:1rem; /* same with indentation */
  height:0;
  border-top:1px solid;
  margin-top:-1px; /* border top width */
  position:absolute;
  top:1rem; /* (line-height/2) */
  left:0;
}


:local .tree li:last-child:before {
  background:white; /* same with body background */
  height:auto;
  top:1rem; /* (line-height/2) */
  bottom:0;
}
