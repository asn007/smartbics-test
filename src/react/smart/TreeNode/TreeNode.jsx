import React from 'react';
import PropTypes from 'prop-types';
import EditableContent from '../EditableContent/EditableContent';
import Button from '../../dumb/Button/Button';
import TreeNodeCreator from '../TreeNodeCreator/TreeNodeCreator';
import style from './TreeNode.css';

export default class TreeNode extends React.PureComponent {
  static propTypes = {
    value: PropTypes.string,
    childNodes: PropTypes.array
  };

  static defaultProps = {
    value: null,
    childNodes: []
  };

  updateGlobalTree = value => {
    this.props.updateNode(this.props.pointer.toString(), {
      value,
      childNodes: this.props.childNodes
    });
  };

  createNode = value => {
    if (!!value)
      this.props.createNode(this.props.pointer.toString(), {
        value,
        childNodes: []
      });
  };

  deleteNode = () => this.props.deleteNode(this.props.pointer.toString());

  renderChildren(isMasterNode = true) {
    const { pointer, childNodes } = this.props;
    return childNodes.map((props, index) => {
      const childPointer = pointer.toString() + '.' + index.toString();
      return React.createElement(TreeNode, {
        ...props,
        key: childPointer,
        pointer: childPointer,
        updateNode: this.props.updateNode,
        createNode: this.props.createNode,
        deleteNode: this.props.deleteNode
      });
    });
  }

  renderRoot() {
    const { childNodes } = this.props;
    return (
      <ul className={style.tree}>
        {this.renderChildren(true)}
        <TreeNodeCreator createNode={this.createNode} />
      </ul>
    );
  }

  render() {
    const { value, childNodes, pointer } = this.props;
    if (pointer.length === 2) return this.renderRoot();
    else
      return (
        <li>
          {value
            ? <span>
                <EditableContent
                  onChange={this.updateGlobalTree}
                  value={value}
                />
                <Button onClick={this.deleteNode} danger>Ã—</Button>
              </span>
            : null}
          {
            <ul>
              {this.renderChildren()}
              <TreeNodeCreator createNode={this.createNode} />
            </ul>
          }
        </li>
      );
  }
}
