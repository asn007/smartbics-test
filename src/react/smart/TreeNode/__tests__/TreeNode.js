import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import TreeNode from '../TreeNode';
import tree from '../../../../helpers/demoTree';



describe('TreeNode', () => {
  test('it should render its markup correctly', () => {
    const component = renderer.create(<TreeNode {...tree} pointer=".0" />);
    expect(component.toJSON()).toMatchSnapshot();
  });
})
