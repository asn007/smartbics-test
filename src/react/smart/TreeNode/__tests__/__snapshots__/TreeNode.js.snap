// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeNode it should render its markup correctly 1`] = `
<ul
  className="tree"
>
  <li>
    <span>
      <div
        className="editableContent"
      >
        <span
          onClick={[Function]}
        >
          Element #1
        </span>
      </div>
      <button
        className="btn  danger "
        onClick={[Function]}
      >
        <div>
          ×
        </div>
      </button>
    </span>
    <ul>
      <li
        className="treeNodeCreator"
      >
        <div
          className="arrow"
          onClick={[Function]}
        >
          →
        </div>
      </li>
    </ul>
  </li>
  <li>
    <span>
      <div
        className="editableContent"
      >
        <span
          onClick={[Function]}
        >
          Element #2
        </span>
      </div>
      <button
        className="btn  danger "
        onClick={[Function]}
      >
        <div>
          ×
        </div>
      </button>
    </span>
    <ul>
      <li
        className="treeNodeCreator"
      >
        <div
          className="arrow"
          onClick={[Function]}
        >
          →
        </div>
      </li>
    </ul>
  </li>
  <li>
    <span>
      <div
        className="editableContent"
      >
        <span
          onClick={[Function]}
        >
          Element #3
        </span>
      </div>
      <button
        className="btn  danger "
        onClick={[Function]}
      >
        <div>
          ×
        </div>
      </button>
    </span>
    <ul>
      <li>
        <span>
          <div
            className="editableContent"
          >
            <span
              onClick={[Function]}
            >
              Child #1 of element #3
            </span>
          </div>
          <button
            className="btn  danger "
            onClick={[Function]}
          >
            <div>
              ×
            </div>
          </button>
        </span>
        <ul>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Another subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Another subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Last subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li
            className="treeNodeCreator"
          >
            <div
              className="arrow"
              onClick={[Function]}
            >
              →
            </div>
          </li>
        </ul>
      </li>
      <li>
        <span>
          <div
            className="editableContent"
          >
            <span
              onClick={[Function]}
            >
              Child #2 of element #3
            </span>
          </div>
          <button
            className="btn  danger "
            onClick={[Function]}
          >
            <div>
              ×
            </div>
          </button>
        </span>
        <ul>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li
            className="treeNodeCreator"
          >
            <div
              className="arrow"
              onClick={[Function]}
            >
              →
            </div>
          </li>
        </ul>
      </li>
      <li>
        <span>
          <div
            className="editableContent"
          >
            <span
              onClick={[Function]}
            >
              Child #3 of element #3
            </span>
          </div>
          <button
            className="btn  danger "
            onClick={[Function]}
          >
            <div>
              ×
            </div>
          </button>
        </span>
        <ul>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li>
            <span>
              <div
                className="editableContent"
              >
                <span
                  onClick={[Function]}
                >
                  Subchild
                </span>
              </div>
              <button
                className="btn  danger "
                onClick={[Function]}
              >
                <div>
                  ×
                </div>
              </button>
            </span>
            <ul>
              <li
                className="treeNodeCreator"
              >
                <div
                  className="arrow"
                  onClick={[Function]}
                >
                  →
                </div>
              </li>
            </ul>
          </li>
          <li
            className="treeNodeCreator"
          >
            <div
              className="arrow"
              onClick={[Function]}
            >
              →
            </div>
          </li>
        </ul>
      </li>
      <li
        className="treeNodeCreator"
      >
        <div
          className="arrow"
          onClick={[Function]}
        >
          →
        </div>
      </li>
    </ul>
  </li>
  <li
    className="treeNodeCreator"
  >
    <div
      className="arrow"
      onClick={[Function]}
    >
      →
    </div>
  </li>
</ul>
`;
